File: Project4/Mario.java
import java.util.ArrayList;
import java.util.Iterator;
import java.awt.Image;
import java.awt.Graphics;

public class Mario extends Sprite {
	int imgNumber;
	int moveSpeed;
	int flightTime;
	int stepSize;
	int sinceStep;
  static Image[] images;

	Mario() {
		super(150, 700, 60, 95, 0, 0);
		imgNumber = 0;
		moveSpeed = 8;
		flightTime = 0;
		isMario = true;
		stepSize = 2;
		loadImages();
	};

	private void loadImages () {
		if (images == null) {
			images = new Image[10];
			for (int i = 0; i <= 9; i++) {
				String imgSrc = "imgs/mario" + (i+1) + ".png";
				images[i] = View.loadImage(imgSrc);
			}
		}
	}

	public void update(Model m) {
		// apply gravity
		applyGravity();
		addVel();

		// control picture
		if (vx != 0) {
			sinceStep++;
			if (sinceStep >= stepSize) {
				imgNumber = (imgNumber+1)%5;
				sinceStep = 0;
			}
			if (vx > 0) {
				imgNumber = imgNumber%5;
			} else if (vx < 0) {
				imgNumber = (imgNumber%5) + 5;
			}
		}

		// control jumping
		flightTime++;

		// check for collision
		Iterator<Sprite> spriterator = m.sprites.iterator();
		while(spriterator.hasNext()) {
			Sprite s = spriterator.next();
			if (isColliding(s)) {
				if (s.isMario)	continue;

				s.spriteHit(m, this);

				int dir = getOut(s);
				if (dir == 1 ) { //on ground
					flightTime = 0;
					vy = 0;
				} else if (dir == 2) {
					vx = 0;
				} else if (dir == 3) {
					vy = 0;
				} else if (dir == 4) {
					vx = 0;
				}
			}
		}
	}

	public Json marshall () {
		Json ob = Json.newObject();
		ob.add("x", x);
		ob.add("y", y);
		ob.add("w", w);
		ob.add("h", h);
		ob.add("type", "mario");
		return ob;
	}

	public void draw(Graphics g, Model m) {
		g.drawImage(images[imgNumber], x - m.camX, y - m.camY, null);
	}

	public void moveX(int xDir) {
		vx = moveSpeed * xDir;
	}

	public void jump() {
		if (flightTime <= 5) {
			vy += -6;
		}
	}

	public void crouch() {

	}

	@Override
	public void spriteHit(Model m, Sprite s) {}
}

File: src/mario.js
class Mario extends Movable {
  constructor(view) {
    super(0,sh - 200,60,95,0,0);
    this.isMario = true;
    this.moveSpeed = 8;
    this.stepSize = 2;
    this.image;
    this.imgNumber = 0;
    this.flightTime = 0;
    this.sinceStep = 0;
    this.jumpCount = 0;
    this.images = []
    this.loadImages();
  }

  loadImages() {
    for (let i = 1; i <= 10; i++) {
      const img = new Image();
      img.src = "imgs/mario" + i + ".png";
      this.images.push(img);
    }
  }

  update(model) {
    this.applyGravity();
    this.addVel();

    this.flightTime++;

    this.sinceStep += Math.abs(this.vx);
    if (this.sinceStep > 20) {
      this.imgNumber = (this.imgNumber+1)%5;
      this.sinceStep = 0;
    }

    if (this.vx < 0) {
      this.imgNumber = this.imgNumber%5 + 5;
    }


    for (const s of model.sprites) {
      if (s == this) continue;
      if (this.isColliding(s)) {
        const dir = this.getOut(s);
				s.spriteHit(model, this);

				if (dir == 1 ) { //on ground
					this.flightTime = 0;
					this.vy = 0;
				} else if (dir == 2) {
					this.vx = 0;
				} else if (dir == 3) {
					this.vy = 0;
				} else if (dir == 4) {
					this.vx = 0;
				}
      }
    }
  }

  draw(paint, model) {
    paint.drawImage(this.images[this.imgNumber], this.x - model.camx, this.y)
  }

  jump() {
    if (this.flightTime <= 5) {
      this.vy += -1;
    }
  }
}